arch/risc-v/src/mpfs/mpfs_coremmc.c

```diff
@@ -0,0 +1,2406 @@
+        return -EINVAL;
+    }
+
+  /* Wait for the response (or timeout) */
+
+  do
+    {
+      status = getreg8(MPFS_COREMMC_ISR);
+    }
+  while (!(status & (COREMMC_ISR_RRI | COREMMC_ISR_ERROR))
+         && --timeout);
+
+  if (timeout == 0 || (status & COREMMC_ISR_ERROR))
+    {
+      mcerr("ERROR: Timeout cmd: %08" PRIx32 " stat: %02" PRIx8 "\n", cmd,
+            status);
+      return -EBUSY;
+    }
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: mpfs_check_recverror
+ *
+ * Description:
+ *   Receive response to SDIO command.
+ *
+ * Input Parameters:
+ *   priv    - Instance of the CoreMMC private state structure.
+ *
+ * Returned Value:
+ *   OK on success; a negated errno if error detected.
+ *
+ ****************************************************************************/
+
+static int mpfs_check_recverror(struct mpfs_dev_s *priv)
+{
+  uint32_t timeout = COREMMC_CMDTIMEOUT;
+  uint8_t regval;
+  int ret = OK;
+
+  regval = getreg8(MPFS_COREMMC_ISR);
+
+  if (regval & COREMMC_ISR_ERROR)
+    {
+      if (regval & COREMMC_ISR_SBI)
+        {
+          mcerr("ERROR: Command timeout: %02" PRIx8 "\n", regval);
+          ret = -ETIMEDOUT;
+        }
+      else if (regval & COREMMC_ISR_TBI)
+        {
+          mcerr("ERROR: Stop bit error: %02" PRIx8 "\n", regval);
+          ret = -EIO;
+        }
+      else
+        {
+          mcerr("ERROR: %02" PRIx8 "\n", regval);
+          ret = -EIO;
+        }
+    }
+  else if (!(regval & COREMMC_ISR_RRI))
+    {
+      mcerr("ERROR: Command not completed: %02" PRIx8 "\n", regval);
+      ret = -EIO;
+    }
+
+  /* Make sure response data ready is set (RDRE) */
+
+  do
+    {
+      regval = getreg8(MPFS_COREMMC_SR);
+    }
+  while (!(regval & (MPFS_COREMMC_SR_RDRE))
+         && --timeout);
+
+  if (timeout == 0)
+    {
+      mcerr("ERROR: Response not ready: %02" PRIx8 "\n", regval);
+      ret = -ETIMEDOUT;
+    }
+
+  if (ret)
+    {
+      /* With an error, reset DAT and CMD lines. Otherwise the next command
+       * will fail as well.
+       */
+
+      mpfs_reset_lines(priv);
+
+      /* Clear all status interrupts */
+
+      mpfs_putreg8(priv, COREMMC_ALL_ICR, MPFS_COREMMC_ICR);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: mpfs_recvshortcrc
+ *
+ * Description:
+ *   Receive response to SDIO command.
+ *
+ * Input Parameters:
+ *   dev    - An instance of the SDIO device interface
+ *   cmd    - Command
+ *   rshort - Buffer for reveiving the response
+ *
+ * Returned Value:
+ *   OK on success; a negated errno on failure.
+ *
+ ****************************************************************************/
+
+static int mpfs_recvshortcrc(struct sdio_dev_s *dev, uint32_t cmd,
+                              uint32_t *rshort)
+{
+  struct mpfs_dev_s *priv = (struct mpfs_dev_s *)dev;
+  uint32_t response = 0;
```

Comment:
minor/optional: visibility can be reduced by declaring in `if (rshort)` scope and removing zero init by:
```
uint32_t response;

response  = getreg8(MPFS_COREMMC_RR1) << 24;
...
```